### PR TITLE
Reservation create [BBS-84]

### PR DESCRIPTION
**Pull request by tripox**

### CODE REVIEW

**Comment by kasperg** [comment]
@tripox: Scrutinizer issues seem relevant here.

File: modules/aleph/aleph.module [hunk 1]
```diff
@@ -11,6 +11,9 @@ use Drupal\aleph\Aleph\Handler\AlephPatronHandler;
 include_once 'aleph.features.inc';
 require __DIR__ . '/vendor/autoload.php';
 
+// Define date formats used by Alma.
+define('ALEPH_DATE', 'Y-m-d');
```
Comment (kasperg) [inline modules/aleph/aleph.module]: A bit of copy/paste in the comment here? 😉

File: modules/aleph/includes/aleph.reservation.inc [hunk 2]
```diff
@@ -213,8 +215,57 @@ function aleph_reservation_list($account, $requested_type = NULL) {
...
+  $reservation = new AlephReservation();
+  $request = new AlephRequest();
+  $request->setRequestDate(new DateTime());
+  $request->setEndRequestDate($end_request_date);
+  $request->setPickupLocation($defaults['aleph_preferred_branch']);
+  $reservation->setRequest($request);
+
+  // Try to make the reservation.
+  $patron = $patron_handler->getPatron();
+  $patron_handler->createReservation($patron, $reservation);
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc]: [Other](https://github.com/ding2/ding2/blob/master/modules/fbs/includes/fbs.reservation.inc#L117) [implementations](https://github.com/ding2/ding2/blob/master/modules/alma/includes/alma.reservation.inc#L304) of `reservation_create` have a return value. This does not. Is that intentional?

File: modules/aleph/src/Aleph/AlephClient.php [hunk 3]
```diff
@@ -278,6 +279,49 @@ public function getReservations(AlephPatron $patron) {
...
+    $rid = $this->mainLibrary . $request->getDocNumber();
+
+    // ADM library code + the item record key.
+    // For example, USM50000238843000320.
+    $iid = $request->getInstitutionCode() . $request->getDocNumber() .
+      $request->getItemSequence();
+
+    return $this->requestRest(
+      'PUT',
+      'patron/' . $patron->getId() . '/record/' . $rid . '/items' . $iid . '/hold',
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: Is there a missing `/` between `'/items'` and `$iid`?

File: modules/aleph/src/Aleph/Entity/AlephRequest.php [hunk 4]
```diff
@@ -249,10 +257,10 @@ public function getRequestType() {
   /**
    * Set the request date.
    *
-   * @param string $request_date
+   * @param \DateTime $request_date
    */
   public function setRequestDate($request_date) {
-    $this->requestDate = $request_date;
+    $this->requestDate = $request_date->format('Y-m-d');
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Entity/AlephRequest.php]: How about using `ALEPH_DATE` here?

File: modules/aleph/src/Aleph/Entity/AlephRequest.php [hunk 5]
```diff
@@ -267,10 +275,10 @@ public function getRequestDate() {
   /**
    * Set the end request date.
    *
-   * @param string $end_request_date
+   * @param \DateTime $end_request_date
    */
   public function setEndRequestDate($end_request_date) {
-    $this->endRequestDate = $end_request_date;
+    $this->endRequestDate = $end_request_date->format('Y-m-d');
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Entity/AlephRequest.php]: As above.

File: modules/aleph/src/Aleph/Entity/AlephRequest.php (hunk 4 above)
Comment (kasperg) [reply modules/aleph/src/Aleph/Entity/AlephRequest.php]: It should also be OK to specify a proper interface as the parameter type e.g.

```php
public function setRequestDate(\DateTimeInterface $request_date) {
```

The interface is preferable to the class so that both DateTime and DateTimeImmutable can be type hinted for.

**Comment by tripox** [comment]
@kasperg, yup I'm on it. Needs some more work and not quite ready yet.
I'll ping you, when it is.

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php [hunk 6]
```diff
@@ -146,10 +147,12 @@ public function getReservations() {
         $request = new AlephRequest();
         $material = new AlephMaterial();
 
+        $end_request_date = (string) $hold_request->xpath('z37/z37-end-request-date')[0];
+
         $request->setStatus((string) $hold_request->xpath('z37/z37-status')[0]);
         $request->setPickupLocation((string) $hold_request->xpath('z37/z37-pickup-location')[0]);
         $request->setOpenDate((string) $hold_request->xpath('z37/z37-open-date')[0]);
-        $request->setEndRequestDate((string) $hold_request->xpath('z37/z37-end-request-date')[0]);
+        $request->setEndRequestDate(DateTime::createFromFormat('Ymd', $end_request_date));
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: Should this not be `ALEPH_DATE`?

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php [hunk 7]
```diff
@@ -197,6 +202,18 @@ public function renewLoans($ids) {
...
 
   /**
+   * Create a reservation for a patron.
+   *
+   * @param AlephPatron $patron
+   * @param AlephReservation $reservation
+   *
+   * @throws \RuntimeException
+   */
+  public function createReservation($patron, $reservation) {
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: How about adding proper classes here for the arguments e.g.

```php
public function createReservation(AlephPatron $patron, AlephReservation $reservation) {

File: modules/aleph/src/Aleph/AlephClient.php [hunk 8]
```diff
@@ -278,6 +279,49 @@ public function getReservations(AlephPatron $patron) {
   }
 
   /**
+   * Create a reservation.
+   *
+   * @internal $rid
+   *    Record ID
+   * @internal $iid
+   *    Item ID
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: Why are these added here?

As I see it the comments do not add much value.

File: modules/aleph/includes/aleph.reservation.inc [hunk 9]
```diff
@@ -213,8 +215,57 @@ function aleph_reservation_list($account, $requested_type = NULL) {
...
+  $patron_handler = new AlephPatronHandler(aleph_client());
+  $patron_handler->setPatron(aleph_get_patron_by_account());
+
+  // Check if the users has this reservation and throw exception.
+  $reservation_exists = FALSE;
+  $reservations = $patron_handler->getReservations();
+
+  foreach ($reservations as $reservation) {
+    if ($id === $reservation->getItem()->getId()) {
+      $reservation_exists = TRUE;
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc]: Why not just throw the exception here instead of waiting for the loop to complete?

File: modules/aleph/includes/aleph.reservation.inc (hunk 2 above)
Comment (kasperg) [reply modules/aleph/includes/aleph.reservation.inc]: As far as I can tell `createReservation` can throw a `\RuntimeException`. That does not seem to be declared here?

File: modules/aleph/includes/aleph.reservation.inc [hunk 10]
```diff
@@ -213,8 +215,57 @@ function aleph_reservation_list($account, $requested_type = NULL) {
...
+ *   Drupal user account object.
+ * @param string $id
+ *   ID of reservation.
+ *
+ * @return array
+ *   Reservation information.
+ *
+ * @throws DingProviderReservationExists
+ * @throws DingProviderReservationNotAllowed
+ * @throws DingProviderReservationNotAvailable
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc]: Not all the exceptions thrown here seems to be thrown in practice e.g. `DingProviderReservationNotAllowed` and `DingProviderReservationNotAvailable`. Is that intentional?

File: modules/aleph/includes/aleph.reservation.inc [hunk 11]
```diff
@@ -213,8 +215,57 @@ function aleph_reservation_list($account, $requested_type = NULL) {
...
+  foreach ($reservations as $reservation) {
+    if ($id === $reservation->getItem()->getId()) {
+      $reservation_exists = TRUE;
+    }
+  }
+
+  if ($reservation_exists) {
+    throw new DingProviderReservationExists();
+  }
+
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc]: [I spotted some discussion about handling of reservations for items already on the shelf](https://reload.atlassian.net/secure/RapidBoard.jspa?rapidView=297&view=detail&selectedIssue=BBS-84). That does not seem to be handled here. Is that intentional?

File: modules/aleph/aleph.module (hunk 1 above)
Comment (kasperg) [reply modules/aleph/aleph.module]: Also `ALEPH_DATE_FORMAT` might be a more appropriate name here.

**Review by kasperg — CHANGES_REQUESTED** [review]
Reviewed. Lots of questions and comments here and there.

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php (hunk 6 above)
Comment (tripox) [reply modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: Nope, this is the format we create the object from.
It is provided as `Ymd` and should be converted to `Y-m-d`.

File: modules/aleph/includes/aleph.reservation.inc (hunk 2 above)
Comment (tripox) [reply modules/aleph/includes/aleph.reservation.inc]: Not necessarily a `\RuntimeException` however, you're right the return statement was missing.

File: modules/aleph/includes/aleph.reservation.inc (hunk 11 above)
Comment (tripox) [reply modules/aleph/includes/aleph.reservation.inc]: getReplyCode would return `0025` in that case.
`$response->success()` would return `FALSE` and an exception is thrown with the error message: "Item or like copy is on shelf. Cannot place hold request."

File: modules/aleph/src/Aleph/Entity/AlephRequestResponse.php [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    $this->note = $note;
+  }
+
+  /**
+   * @return bool
+   */
+  public function success() {
+    if ($this->getReplyCode() === '0000') {
+      return TRUE;
+    }
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Entity/AlephRequestResponse.php]: This should probably be `return $this->getReplyCode() === '0000'`

File: modules/aleph/src/Aleph/Entity/AlephRequestResponse.php [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+ *
+ * This entity describes the response from Aleph.
+ *
+ * @package Drupal\aleph\Aleph\Entity
+ */
+class AlephRequestResponse {
+
+  /**
+   * @var $replyCode
+   *  The reply code is 0000 on success.
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Entity/AlephRequestResponse.php]: It is probably worth showing that the reply code is a string - not a integer.

If we have a link to a description of these reply codes and other aspects of the aleph response then adding it would be appropriate.

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php (hunk 6 above)
Comment (kasperg) [reply modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: So we send dates to Aleph in one format `Y-m-d` and recieve them in another `Ymd`?

File: modules/aleph/includes/aleph.reservation.inc [hunk 14]
```diff
@@ -213,8 +215,67 @@ function aleph_reservation_list($account, $requested_type = NULL) {
...
+    $response = $patron_handler->createReservation($patron, $reservation);
+  }
+  catch (Exception $e) {
+    watchdog_exception('aleph', $e);
+  }
+
+  if ($response->success()) {
+    return array(
+      'branch' => $request->getPickupLocation(),
+      'queue_number' => NULL,
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc]: Adding a comment about why `queue_number` is `NULL` would be nice. Probably because Aleph does not support queue numbers?

File: modules/aleph/includes/aleph.reservation.inc [hunk 15]
```diff
@@ -213,8 +215,67 @@ function aleph_reservation_list($account, $requested_type = NULL) {
...
+  $patron = $patron_handler->getPatron();
+  $response = NULL;
+  try {
+    $response = $patron_handler->createReservation($patron, $reservation);
+  }
+  catch (Exception $e) {
+    watchdog_exception('aleph', $e);
+  }
+
+  if ($response->success()) {
```
Comment (kasperg) [inline modules/aleph/includes/aleph.reservation.inc]: If an exception occurs for `createReservation()` then `$response` will be `NULL` and you will get an error trying to call a method on a non-object.

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php [hunk 16]
```diff
@@ -197,6 +203,26 @@ public function renewLoans($ids) {
...
+   * @return \Drupal\aleph\Aleph\Entity\AlephRequestResponse
+   */
+  public function createReservation($patron, $reservation) {
+    $response = $this->client->createReservation($patron,
+      $reservation->getRequest());
+
+    $request_response = new AlephRequestResponse();
+    $request_response::createRequestResponseFromXML($response);
+
+    return $request_response;
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: This does not make sense to me. `createRequestResponseFromXML` is a static method with a return value. This tries to call it from an object instance but do not assign the return value.

This should be `return AlephRequestResponse::createRequestResponseFromXML($response);`

**Review by kasperg — CHANGES_REQUESTED** [review]

**Review by kasperg — APPROVED** [review]

**Comment by kasperg** [comment]
:+1: LGTM
